napari/layers/base/base.py

```diff
@@ -205,44 +206,25 @@ def __init__(
         #   than the maximum allowed texture size of your graphics card.
         # 2. `data2world`: The main transform mapping data to a world-like
         #   coordinate.
-        # 3. `world2grid`: An additional transform mapping world-coordinates
+        # 3. `world2world`: An extra transform applied in world-coordinates that
+        #   typically aligns this layer with another.
+        # 4. `world2grid`: An additional transform mapping world-coordinates
         #   into a grid for looking at layers side-by-side.
-
-        # First create the `data2world` transform from the input parameters
-        if affine is None:
-            if scale is None:
-                scale = [1] * ndim
-            if translate is None:
-                translate = [0] * ndim
-            data2world_transform = Affine(
-                scale,
-                translate,
-                rotate=rotate,
-                shear=shear,
-                name='data2world',
-            )
-        elif isinstance(affine, np.ndarray) or isinstance(affine, list):
-            data2world_transform = Affine(
-                affine_matrix=np.array(affine),
-                name='data2world',
-            )
-        elif isinstance(affine, Affine):
-            affine.name = 'data2world'
-            data2world_transform = affine
-        else:
-            raise TypeError(
-                trans._(
-                    'affine input not recognized. must be either napari.utils.transforms.Affine, ndarray, or None. Got {dtype}',
-                    deferred=True,
-                    dtype=type(affine),
-                )
-            )
-
+        data2world_transform = CompositeAffine(
+            self.ndim,
+            scale=scale,
+            translate=translate,
+            rotate=rotate,
+            shear=shear,
+            name='data2world',
+        )
+        world2world_transform = self._coerce_affine(affine)
         self._transforms = TransformChain(
```

Comment:
the syntax of accessing items in `self._transforms` has always confused me.  It's a mutable sequence (that also happens to be indexable with string names... which makes sense for our public viewer.layers API but is a bit strange here).  but transforms seem to be accessed either directly via name, or via specific indices (e.g. `data_corners = self._transforms[1:].simplified.inverse(corner_pixels)`) ... both of those kind of make it seem like this really shouldn't be a mutable list at all, perhaps a named tuple?  Is that true?  Do we need to support more/less than 4 specific transforms on layers?  and do they need to be re-arrangeble?

seems like now might be a good time to rethink that if you have any insights while working on it.

Reply:
Agreed that `_transforms` does not need to be a mutable collection based on current usage within napari. I would even consider making each element immutable (at least from the public interface) given that transforms are often pretty cheap to create (and changing a property often requires recreating most of the transform anyway).

More generally (and still with respect to napari's internal usage) I think that the `transforms` package is a little over-engineered with `TransformChain` being an example of that. Given that Layer ends up defining 3 (now 4) matrix transforms, I think it would be simpler to just define those as hidden attributes and expose properties that multiply them in certain orders (e.g. like `data2world_transform` here). It would be good to know if anyone depends on some of the more advanced behavior of the `transforms` package to understand if that's something we might want to simplify later.

The base `Transform` and `TransformChain` classes could have some interesting potential if we want to support non-matrix based transform in the future (e.g. supporting transforms defined by displacement fields), but it's probably worth thinking harder about that only when we have a need to support it.

Reply:
I can see multiple use cases for directly supporting displacement fields in my field (cryo-electron microscopy)

Specifically, there is a [recent paper](https://www.biorxiv.org/content/10.1101/2021.04.22.440893v1.full.pdf) which demonstrate the ability to learn a 3D structure and corresponding displacement fields which describe molecular motion from data.

We also routinely perform non-rigid image registration (e.g. [MotionCor2](https://www.nature.com/articles/nmeth.4193) and only tend to visualise the registered and summed result

In both cases, the ability to work with the raw data transformed by the displacement fields could open up lots of new visualisation possibilities :)